components/brave_wallet/browser/solana_transaction.cc

```diff
@@ -176,35 +182,78 @@ bool SolanaTransaction::operator!=(const SolanaTransaction& tx) const {
   return !operator==(tx);
 }
 
+// Get serialized message bytes and array of signers.
+// Serialized message will be the result of decoding
+// sign_tx_param_->encoded_serialized_msg if exists.
+absl::optional<std::pair<std::vector<uint8_t>, std::vector<std::string>>>
+SolanaTransaction::GetSerializedMessage() const {
+  std::vector<std::string> signers;
+  auto message_bytes = message_.Serialize(&signers);
+  if (!message_bytes || signers.empty())
+    return absl::nullopt;
+
+  // If sign_tx_param_ exists, decode encoded_serialized_msg from dApp to be
+  // the serialized message byte array.
+  if (sign_tx_param_ &&
+      !Base58Decode(sign_tx_param_->encoded_serialized_msg,
+                    &message_bytes.value(), kSolanaMaxTxSize, false)) {
+    return absl::nullopt;
+  }
+
+  return std::make_pair(std::move(*message_bytes), std::move(signers));
+}
+
 // Get serialized and signed transaction.
 // A transaction contains a compact-array of signatures, followed by a message.
 // A compact-array is serialized as the array length, followed by each array
 // item. The array length is a special multi-byte encoding called compact-u16.
 // See https://docs.solana.com/developing/programming-model/transactions.
 absl::optional<std::vector<uint8_t>>
 SolanaTransaction::GetSignedTransactionBytes(
-    KeyringService* keyring_service) const {
-  if (!keyring_service)
+    KeyringService* keyring_service,
+    std::vector<uint8_t>* fee_payer_signature) const {
+  if (!keyring_service && !fee_payer_signature)
     return absl::nullopt;
 
-  std::vector<std::string> signers;
-  auto message_bytes = message_.Serialize(&signers);
-  if (!message_bytes || signers.empty())
+  auto message_signers_pair = GetSerializedMessage();
+  if (!message_signers_pair)
     return absl::nullopt;
+  auto& message_bytes = message_signers_pair->first;
+  auto& signers = message_signers_pair->second;
 
   std::vector<uint8_t> transaction_bytes;
   // Compact array of signatures.
   CompactU16Encode(signers.size(), &transaction_bytes);
```

Comment:
This should be exactly same size as the total signatures. If total signatures is less than this size, when a dApp tries to use byte array to construct a web3 Transaction object, that operation might fail.

Reply:
Also in https://docs.solana.com/developing/programming-model/transactions#transaction-format
`t also verifies that each signature was signed by the private key corresponding to the public key at the same index in the message's account addresses array`
So we should follow the order of signer in message.

Reply:
Use actual size of total signatures: https://github.com/brave/brave-core/pull/14001/commits/30f3b6d3cc5e2822bdea392e9d4deb1cad59ecb4
Fix order of signer: https://github.com/brave/brave-core/pull/14001/commits/4ee63086cdc0d66ea8b7845fceb6eecead0567ce